ydb/src/credentials.rs

```diff
@@ -147,6 +151,163 @@ impl Credentials for CommandLineYcToken {
+
+        Ok(Self {
+            key_id: key.id,
+            service_account_id: key.service_account_id,
+            private_key: SecretString::new(key.private_key),
+        })
+    }
+
+    const AUDIENCE: &'static str = "https://iam.api.cloud.yandex.net/iam/v1/tokens";
+    const JWT_TOKEN_LIFE_TIME: usize = 3600;
+
+    fn build_jwt(&self) -> YdbResult<String> {
+        let private_key = self.private_key.expose_secret().as_bytes();
+
+        #[derive(Debug, Serialize, Deserialize)]
+        struct Claims {
+            aud: String, // Optional. Audience
+            exp: usize, // Required (validate_exp defaults to true in validation). Expiration time (as UTC timestamp)
+            iat: usize, // Optional. Issued at (as UTC timestamp)
+            iss: String, // Optional. Issuer
+        }
+
+        let iat = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Time went backwards")
+            .as_secs() as usize;
+
+        let mut header = Header::new(Algorithm::PS256);
+        header.kid = Some(self.key_id.clone());
+        header.alg = Algorithm::PS256;
+        header.typ = Some("JWT".to_string());
+
+        let claims = Claims {
+            exp: iat + Self::JWT_TOKEN_LIFE_TIME,
+            aud: Self::AUDIENCE.to_owned(),
+            iat,
+            iss: self.service_account_id.clone(),
+        };
+        let token = encode(
+            &header,
+            &claims,
+            &EncodingKey::from_rsa_pem(private_key).map_err(|e| YdbError::custom(e.to_string()))?,
+        )
+        .map_err(|e| YdbError::custom(format!("can't build jwt: {}", e)))?;
+
+        debug!("Token was built");
+        Ok(token)
+    }
+
+    fn time_to_instant(time: chrono::DateTime<chrono::Utc>) -> Instant {
+        let duration = time - chrono::Utc::now();
+
+        Instant::now()
+            .checked_add(duration.to_std().unwrap())
+            .expect("Failed to convert Duration to Instant")
+    }
+}
+
+impl Credentials for ServiceAccountCredentials {
+    fn create_token(&self) -> YdbResult<TokenInfo> {
+        const API_URL: &'static str = "https://iam.api.cloud.yandex.net/iam/v1/tokens";
+        use chrono::Utc;
+        #[derive(Deserialize)]
+        struct TokenResponse {
+            #[serde(rename = "iamToken")]
+            iam_token: String,
+            #[serde(rename = "expiresAt")]
+            expires_at: DateTime<Utc>,
+        }
+
+        #[derive(Serialize)]
+        struct TokenRequest {
+            jwt: String,
+        }
+
+        let jwt = self.build_jwt()?;
+
+        let req = TokenRequest { jwt };
+        let client = reqwest::blocking::Client::new();
+        let res: TokenResponse = client
+            .request(reqwest::Method::POST, API_URL)
+            .json(&req)
+            .send()?
+            .json()?;
+
+        Ok(TokenInfo::token(format!("Bearer {}", res.iam_token))
+            .with_renew(Self::time_to_instant(res.expires_at)))
```

Comment:
IAM has recommend for [renew token before it expires](https://yandex.cloud/en/docs/iam/concepts/authorization/iam-token).

It allow to prevent problems:
1. Time skew between IAM and client system.
2. Additional latency for wait new certificate
3. Continue to work if iam service failed on renew certificate, while IAM team repairing the service.

By common rule: we request IAM certificate for 11-12 hours and renew the token while it has about 90% of lifetime (in that case - after one hour of token usage).

Reply:
Ok. Am I required to replace the original token lifetime value with 720 instead of 3600? in create_token(), I return by analogy with GCEMetadata. Then, somewhere in the code, a decision is made about 90%?